sklearn/cross_validation.py

```diff
@@ -332,6 +332,109 @@ def __len__(self):
+    ----------
+    y: iterable of shape (n_samples, )
+        contains an id for each sample
+        The folds are built so that the same id doesn't appear in two different folds
+    
+    n_folds: int, default is 3
+        number of folds to split the data into
+        
+    Returns
+    -------
+    folds: numpy array of shape (n_samples, )
+        array of integers between 0 and (n_folds - 1)
+        folds[i] contains the folds to which sample i is assigned.
+        
+    Notes
+    -----
+    The folds are built by distributing the labels by frequency of appearance.
+    """
+    labels = np.array(y)
+    unique_labels = np.unique(labels)
+    
+    # number of occurrence of each label (its "weight")
+    weight_per_label = sorted([(sum(labels == label), label) for label in unique_labels])
+    # Total weight of each fold
+    weight_per_fold = np.zeros(n_folds)
+    # For each sample, a digit between 0 and (n_folds - 1) to tell which fold it belongs to
+    folds = np.zeros(len(labels))
+    
+    # While there are weights, distribute them
+    # Specifically, add the biggest weight to the lightest fold
+    while weight_per_label:
+        ind_min = np.argmin(weight_per_fold)
+        w, label = weight_per_label.pop()
+        weight_per_fold[ind_min] += w
+        folds[labels == label] = ind_min
+    
+    return folds
+
+
+class DisjointLabelKfold(_BaseKFold):
+    def __init__(self, y, n_folds=3):
+        """Creates K approximately equilibrated folds
+            where the same label will not appear in two different folds
+
+        Parameters
+        ----------
+        y : array-like with shape (n_samples, )
+            Contains a label for each sample.
+            The folds are built so that the same label doesn't appear in two different folds.
+
+        n_folds : int, default is 3
+            Number of folds.
+
+        Examples
+        --------
+        >>> from sklearn import cross_validation
+        >>> X = np.array([[1, 2], [3, 4], [5, 6], [7, 8]])
+        >>> y = np.array([1, 2, 3, 4])
+        >>> labels = np.array([0, 0, 2, 2])
+        >>> dl_kfold = cross_validation.DisjointLabelKfold(labels, n_folds=2)
+        >>> len(dl_kfold)
+        2
+        >>> print(dl_kfold)
+        sklearn.cross_validation.DisjointLabelKfold(n_labels=4, n_folds=2)
+        >>> for train_index, test_index in dl_kfold:
+        ...     print("TRAIN:", train_index, "TEST:", test_index)
+        ...     X_train, X_test = X[train_index], X[test_index]
+        ...     y_train, y_test = y[train_index], y[test_index]
+        ...     print(X_train, X_test, y_train, y_test)
+        ... 
+        TRAIN: [0 1] TEST: [2 3]
+        [[1 2]
+         [3 4]] [[5 6]
+         [7 8]] [1 2] [3 4]
+        TRAIN: [2 3] TEST: [0 1]
+        [[5 6]
+         [7 8]] [[1 2]
+         [3 4]] [3 4] [1 2]
+        """
+        self.n_folds = n_folds
+        self.n = len(y)
+        self.idxs = disjoint_label_folds(y=y, n_folds=n_folds)
```

Comment:
Perhaps you should also call the super constructor here; it checks, e.g., if the arguments are valid:
```python
super(LabelSegmentedKFold, self).__init__(np.unique(labels).size, n_folds, shuffle, random_state)
```
It expects also ``shuffle`` and ``random_state``. Does it makes sense to extend the functionality in this direction?